src/dex/mod.rs

```diff
@@ -0,0 +1,240 @@
+    /// - Username: `user`
+    /// - Password: `user`
+    /// - User ID: `user`
+    ///
+    /// Users can only be added before the container starts.
+    pub fn with_simple_user(self) -> Self {
+        self.with_user(User::simple_user())
+    }
+
+    /// Appends the specified user.
+    ///
+    /// Users can only be added before the container starts.
+    pub fn with_user(self, user: User) -> Self {
+        Self {
+            users: self.users.into_iter().chain(vec![user]).collect(),
+            ..self
+        }
+    }
+
+    /// Appends a client with
+    /// - Id: `client`
+    /// - Redirect URI: `http://localhost/oidc-callback`
+    /// - Secret: `secret`
+    ///
+    /// Clients can only be added before the container starts.
+    pub fn with_simple_client(self) -> Self {
+        self.with_client(PrivateClient::simple_client())
+    }
+
+    /// Appends the specified client.
+    /// Clients can only be added before the container starts.
+    pub fn with_client(self, client: PrivateClient) -> Self {
+        Self {
+            clients: self.clients.into_iter().chain(vec![client]).collect(),
+            ..self
+        }
+    }
+
+    /// Enables grant_type 'password' (usually for testing purposes)
+    pub fn with_allow_password_grants(self) -> Self {
+        Self {
+            allow_password_grants: true,
+            ..self
+        }
+    }
+}
+
+impl Dex {
+    fn generate_config(&self, host: &str, host_port: u16) -> ExecCommand {
+        let config = config::Config {
+            issuer: format!("http://{}:{}", host, host_port),
+            enable_password_db: true,
+            storage: config::Storage::sqlite(),
+            web: config::Web::http(),
+            static_clients: self.clients.clone(),
+            static_passwords: self.users.clone(),
+            oauth2: if !self.allow_password_grants {
+                None
+            } else {
+                Some(OAuth2::allow_password_grant())
+            },
+        };
+
+        let config = serde_json::to_string(&config)
+            .expect("Parsing should only fail if structs were defined incorrectly.");
+
+        ExecCommand::new(vec![
+            "/bin/sh",
+            "-c",
+            &format!("echo '{}' > {}", config, CONFIG_FILE),
+        ])
+    }
+}
+
+impl Image for Dex {
+    fn name(&self) -> &str {
+        NAME
+    }
+
+    fn tag(&self) -> &str {
+        &self.tag
+    }
+
+    fn ready_conditions(&self) -> Vec<WaitFor> {
+        vec![WaitFor::Http(
+            HttpWaitStrategy::new("/.well-known/openid-configuration")
+                .with_port(HTTP_PORT)
+                .with_expected_status_code(200u16),
+        )]
+    }
+
+    fn cmd(&self) -> impl IntoIterator<Item = impl Into<Cow<'_, str>>> {
+        // Stolen from the Java implementation:
+        // https://github.com/Kehrlann/testcontainers-dex/tree/main/testcontainers-dex/src/main/java/wf/garnier/testcontainers/dexidp/DexContainer.java#L116
```

Comment:
Small rewording and corrected link (if you don't mind)
```suggestion
        // Borrowed from the Java implementation:
        // https://github.com/Kehrlann/testcontainers-dex/blob/00e58fb25c38fe26279e3c8d5fe1fbf9b23f04c0/testcontainers-dex/src/main/java/wf/garnier/testcontainers/dexidp/DexContainer.java#L85-L94
```